files, = glob_wildcards("data/{sample}.fastq.gz")




include: "rules/gbc_variant_call.smk"


#route=config.get("route","bowtie2")



rule all:
	input:
		"results/main/output_snp.filtered.vcf.gz"



""" if route == "bowtie2":
	rule all:
		input:
			expand("results/count_tables/{gene}.tsv",gene=gene),
			expand(["results/count_tables/" + x + ".tsv" for x in ["lncRNA","misc_RNA","protein_coding","snoRNA","snRNA","scaRNA"]]),
			"results/file_statistics.csv",
			expand("analyses/fastqc/{sample}.trimmed.FastQC/{sample}.trimmed_fastqc.zip",sample=files),
			"results/multiqc_report.html",
			"results/mirtrace/mirtrace-report.html"



elif route == "bowtie1":
	print("Bowtie1 route has been selected")
	rule all_bowtie1:
		input:
			expand("analyses/bowtie1_mappings/{sample}.sorted.bam",sample=files)


elif route == "stats" or route == "statistics":
	print("Statistics route has been selected. This workflow requires Bowtie1 outputs")
	rule statistics:
		input:
			expand("analyses/bowtie1_mappings/{sample}.sorted.bam",sample=files)
	

else:
	print("Please select a correct route...")
	print("bowtie1/bowtie2")
 """
